### PR TITLE
NIAD-3213 - Failure to Process DiagnosticReport with a comment-less Filing Comment and Absence of Specimen

### PR DESCRIPTION
**Pull request by martin-nhs**
## What

This PR addresses a bug encountered when processing a `DiagnosticReport` that contains a filing comment without any associated content (comment-less) and lacks a corresponding `Specimen`. The changes in this PR ensure that such cases are handled gracefully without errors. 

## Why

More detail can be found on the [ticket](https://nhse-dsic.atlassian.net/browse/NIAD-3213).

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have updated the [Changelog](CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users

### CODE REVIEW

File: service/src/test/resources/ehr/mapper/diagnosticreport/fhir_bundle.json [hunk 1]
```diff
@@ -288,6 +288,51 @@
...
+          }
+        ],
+        "related":[
+          {
+            "type":"derived-from",
+            "target":{
+              "reference":"Observation/461839A4-9B28-46D4-BC27-297C46F4F80E"
+            }
+          }
+        ]
```
Comment (adrianclay) [inline service/src/test/resources/ehr/mapper/diagnosticreport/fhir_bundle.json:333]: Docs for filing comment related:
> Reference to the test result or test group header that the filing comments resource relates to. Where the filing comments relate to the test report, the reference is made from the ‘Test report’ to the filing comment only.

I think we should remove this derived-from, to be a bit more realistic to the spec.

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java [hunk 2]
```diff
@@ -248,6 +249,41 @@ void When_DiagnosticReport_With_ObservationEffectivePeriodAndCommentNote_Expect_
...
+                ),
+                randomIdGeneratorService,
+                confidentialityService
+            ),
+            new ParticipantMapper(),
+            randomIdGeneratorService,
+            confidentialityService
+        );
+
+        final String actualXml = mapper.mapDiagnosticReportToCompoundStatement(diagnosticReport);
```
Comment (adrianclay) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java:284]: Probably worth asserting against this.

File: service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-no-specimen.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  ],
+  "result": [
+    {
+      "reference": "Observation/FILING-COMMENT-WITH-NO-COMMENT"
+    },
+    {
+      "reference": "Observation/6E30A9E3-FF9A-4868-8FCD-7DAC26A16E78"
+    }
+  ],
+  "conclusion": "COELIAC"
```
Comment (adrianclay) [inline service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-no-specimen.json:60]: Let's simplify our test, by removing this optional field.

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java [hunk 4]
```diff
@@ -248,6 +249,41 @@ void When_DiagnosticReport_With_ObservationEffectivePeriodAndCommentNote_Expect_
         assertThatXml(actualXml).containsAllXPaths(expectedXPaths);
     }
 
+    @Test
+    void When_DiagnosticReport_With_NoReferencedSpecimenAndFilingCommentWithNoComment_Expect_ValidCompoundStatementOutput() {
```
Comment (adrianclay) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java:253]: I think this "Expect ..." needs tweaking a little bit.

File: service/src/test/resources/ehr/mapper/diagnosticreport/fhir_bundle.json (hunk 1 above)
Comment (martin-nhs) [reply service/src/test/resources/ehr/mapper/diagnosticreport/fhir_bundle.json:333]: Agreed!

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 5]
```diff
@@ -256,7 +256,7 @@ private void buildNarrativeStatementForObservationComments(
         reportLevelNarrativeStatements.append(narrativeStatementObservationComments);
     }
 
-    private boolean hasCommentNote(Observation observation) {
+    static boolean hasCommentNote(Observation observation) {
```
Comment (adrianclay) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:259]: I feel like `isFilingComment` is clearer.

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java [hunk 6]
```diff
@@ -248,6 +250,44 @@ void When_DiagnosticReport_With_ObservationEffectivePeriodAndCommentNote_Expect_
...
 
+    @Test
+    void When_DiagnosticReport_With_NoReferencedSpecimenAndFilingCommentWithNoComment_Expect_MatchesSnapshotXml() {
+        final String diagnosticReportFileName = "diagnostic-report-with-no-specimen.json";
+        final DiagnosticReport diagnosticReport = getDiagnosticReportResourceFromJson(diagnosticReportFileName);
+        final Bundle bundle = getBundleResourceFromJson(INPUT_JSON_BUNDLE);
+        final InputBundle inputBundle = new InputBundle(bundle);
+        final String expectedXml = getXmlStringFromFile(TEST_FILE_DIRECTORY, "diagnostic-report-with-no-specimen.xml");
+
+        when(specimenMapper.mapSpecimenToCompoundStatement(any(), anyList(), any()))
```
Comment (martin-nhs) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java:261]: Let's restrict this a bit more, avoiding the use of `any()`.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/SpecimenMapper.java [hunk 7]
```diff
@@ -141,13 +142,15 @@ private String mapObservationsAssociatedWithSpecimen(Specimen specimen, List<Obs
         if (dummySpecimenOrObservationExists(specimen, observations)) {
             observationsAssociatedWithSpecimen = observations;
         } else {
+            // Implicitly filtering out filing comments.
```
Comment (adrianclay) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/SpecimenMapper.java:145]: Is this comment useful still?

**Review by adrianclay — APPROVED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/SpecimenMapper.java (hunk 7 above)
Comment (martin-nhs) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/SpecimenMapper.java:145]: Agreed, separate PR 😄

**Pull request merged**